lib/active_model_serializers/adapter/json_api/deserialization.rb

```diff
@@ -74,21 +74,21 @@ module Deserialization
         #
         def parse!(document, options = {})
           parse(document, options) do |invalid_payload, reason|
-            fail InvalidDocument, "Invalid payload (#{reason}): #{invalid_payload}"
+            fail JSONAPI::Parser::InvalidDocument, "Invalid payload (#{reason}): #{invalid_payload}"
           end
         end
 
         # Same as parse!, but returns an empty hash instead of raising InvalidDocument
         # on invalid payloads.
         def parse(document, options = {})
-          document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
-
-          validate_payload(document) do |invalid_document, reason|
-            yield invalid_document, reason if block_given?
-            return {}
-          end
+          document = JSONAPI.parse(document, options)
+          document = document.to_hash
```

Comment:
The parser now only validates a document, it does not build an object-like structure for traversing it anymore (which was a perf issue). AMS could possibly make use of [jsonapi-validator](https://github.com/beauby/jsonapi-validator) which does domain-specific validations of payloads, or [jsonapi-deserializable](https://github.com/beauby/jsonapi-deserializable) which does general-purpose deserialization of JSONAPI resource/relationship payloads.